File: prolog/difference_of_squares.pl
square_of_sum(N, Result) :-
    Sum is N * (N+1) / 2,
    Result is Sum * Sum.

sum_of_squares(1, 1).

sum_of_squares(N, Result) :-
    Prev is N - 1,
    sum_of_squares(Prev, X),
    Result is N * N + X.

difference(N, Result) :-
    square_of_sum(N, X),
    sum_of_squares(N, Y),
    Result is X - Y.

File: dart/difference_of_squares.dart
class DifferenceOfSquares {
  int squareOfSum(int n) {
    int sum = List.generate(n, (i) => i + 1).reduce((a, b) => a + b);
    return sum * sum;
  }

  int sumOfSquares(int n) {
    return List.generate(n, (i) => (i + 1) * (i + 1)).reduce((a, b) => a + b);
  }

  int differenceOfSquares(int n) {
    return squareOfSum(n) - sumOfSquares(n);
  }
}
